File: backend/ai_engine/indicators.py
import pandas as pd
import numpy as np

def calculate_rsi(series, period=14):
    delta = series.diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=period, min_periods=1).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=period, min_periods=1).mean()
    rs = gain / (loss.replace(0, 1e-9))
    return 100 - (100 / (1 + rs))

def calculate_macd(series, fast=12, slow=26, signal=9):
    ema_fast = series.ewm(span=fast, adjust=False).mean()
    ema_slow = series.ewm(span=slow, adjust=False).mean()
    macd = ema_fast - ema_slow
    macd_signal = macd.ewm(span=signal, adjust=False).mean()
    macd_hist = macd - macd_signal
    return macd, macd_signal, macd_hist

def calculate_bollinger_bands(series, period=20, std_dev=2):
    sma = series.rolling(window=period, min_periods=1).mean()
    rolling_std = series.rolling(window=period, min_periods=1).std().fillna(0)
    upper_band = sma + (rolling_std * std_dev)
    lower_band = sma - (rolling_std * std_dev)
    return upper_band, sma, lower_band

def calculate_atr(df, period=14):
    high_low = df['high'] - df['low']
    high_close = (df['high'] - df['close'].shift()).abs()
    low_close = (df['low'] - df['close'].shift()).abs()
    tr = pd.concat([high_low, high_close, low_close], axis=1).max(axis=1)
    return tr.rolling(window=period, min_periods=1).mean()

def compute_all_indicators(df):
    """
    Expects a pandas DataFrame with columns: open, high, low, close, volume (sorted by date ascending)
    """
    data = df.copy()
    if 'close' not in data.columns:
        return data

    data['rsi'] = calculate_rsi(data['close'], 14)
    
    macd, macd_signal, macd_hist = calculate_macd(data['close'])
    data['macd'] = macd
    data['macd_signal'] = macd_signal
    data['macd_hist'] = macd_hist
    
    data['sma_20'] = data['close'].rolling(window=20, min_periods=1).mean()
    data['sma_50'] = data['close'].rolling(window=50, min_periods=1).mean()
    
    data['ema_20'] = data['close'].ewm(span=20, adjust=False).mean()
    data['ema_50'] = data['close'].ewm(span=50, adjust=False).mean()
    
    upper, mid, lower = calculate_bollinger_bands(data['close'], 20, 2)
    data['bb_upper'] = upper
    data['bb_middle'] = mid
    data['bb_lower'] = lower
    data['bb_width'] = (upper - lower) / (mid.replace(0, 1e-9))
    
    data['atr'] = calculate_atr(data, 14)
    data['volatility_pct'] = (data['atr'] / data['close']) * 100
    data['momentum_5d'] = ((data['close'] - data['close'].shift(5)) / data['close'].shift(5)) * 100
    
    # Fill any initial NaNs
    data = data.ffill().bfill().fillna(0)
    return data

def indicator_signals(df):
    """
    Interpret the mathematical state of the indicators on the latest bar so the
    UI can label them in plain words (no guessing):
      - RSI state          : OVERSOLD (<30) / OVERBOUGHT (>70) / NEUTRAL
      - MACD               : value, signal, histogram + a recent BULLISH/BEARISH
                             crossover (hist sign flip within the last 6 bars) and
                             whether MACD sits above or below its signal line
      - Trend              : price vs SMA20/50 and the SMA20 vs SMA50 relationship
      - Bollinger position : where price sits inside the bands (lower..upper)
    Expects output of compute_all_indicators (sorted ascending by date).
    """
    if df is None or len(df) < 3:
        return None
    latest = df.iloc[-1]
    close = float(latest.get('close', 0))

    def _f(v):
        try:
            return float(v)
        except (TypeError, ValueError):
            return None

    rsi = _f(latest.get('rsi'))
    rsi_state = None
    if rsi is not None:
        if rsi >= 70:
            rsi_state = "OVERBOUGHT"
        elif rsi <= 30:
            rsi_state = "OVERSOLD"
        else:
            rsi_state = "NEUTRAL"

    macd = _f(latest.get('macd'))
    macd_signal = _f(latest.get('macd_signal'))
    macd_hist = _f(latest.get('macd_hist'))
    macd_cross = None
    macd_state = None
    if macd is not None and macd_signal is not None:
        macd_state = "ABOVE" if macd >= macd_signal else "BELOW"
        hist = df['macd_hist'].tail(6).to_numpy(dtype=float)
        for i in range(len(hist) - 1, 0, -1):
            if hist[i - 1] <= 0 < hist[i]:
                macd_cross = "BULLISH"
                break
            if hist[i - 1] >= 0 > hist[i]:
                macd_cross = "BEARISH"
                break

    sma20 = _f(latest.get('sma_20'))
    sma50 = _f(latest.get('sma_50'))
    ema20 = _f(latest.get('ema_20'))
    ema50 = _f(latest.get('ema_50'))
    vs_sma20 = None
    vs_sma50 = None
    if sma20:
        vs_sma20 = round(((close - sma20) / sma20) * 100.0, 2)
    if sma50:
        vs_sma50 = round(((close - sma50) / sma50) * 100.0, 2)
    if sma20 and sma50:
        sma_trend = "UPTREND" if sma20 > sma50 else ("DOWNTREND" if sma20 < sma50 else "SIDEWAYS")
    else:
        sma_trend = None

    bb_upper = _f(latest.get('bb_upper'))
    bb_lower = _f(latest.get('bb_lower'))
    bb_position = None
    bb_state = None
    if bb_upper is not None and bb_lower is not None and bb_upper > bb_lower:
        bb_position = round(((close - bb_lower) / (bb_upper - bb_lower)) * 100.0, 1)
        if bb_position >= 90:
            bb_state = "AT_UPPER"
        elif bb_position >= 60:
            bb_state = "UPPER_HALF"
        elif bb_position <= 10:
            bb_state = "AT_LOWER"
        else:
            bb_state = "LOWER_HALF"

    return {
        "rsi": round(rsi, 1) if rsi is not None else None,
        "rsiState": rsi_state,
        "macd": round(macd, 4) if macd is not None else None,
        "macdSignal": round(macd_signal, 4) if macd_signal is not None else None,
        "macdHist": round(macd_hist, 4) if macd_hist is not None else None,
        "macdCross": macd_cross,
        "macdState": macd_state,
        "priceVsSma20": vs_sma20,
        "priceVsSma50": vs_sma50,
        "ema20": round(ema20, 2) if ema20 is not None else None,
        "ema50": round(ema50, 2) if ema50 is not None else None,
        "smaTrend": sma_trend,
        "bbPosition": bb_position,
        "bbState": bb_state,
        "atr": round(_f(latest.get('atr')) or 0, 2),
        "volatilityPct": round(_f(latest.get('volatility_pct')) or 0, 2),
        "momentum5d": round(_f(latest.get('momentum_5d')) or 0, 2),
    }

def project_duration(sr, projection, price=None):
    """
    Answer "how many more sessions could this trend last" using real math:
    if the stock is trending, how many sessions does the observed daily trend
    rate need to travel from the current price to resistance (upside) or down
    to support (downside). Distance-to-level is the honest cap on a trend, and
    the regression R^2 tells us how much to trust the trend at all.
    """
    if not sr or not projection or price is None:
        return None
    daily_rate = projection.get('dailyRatePct')
    if daily_rate is None or abs(float(daily_rate)) < 1e-6:
        return None

    resistance = float(sr.get('resistance') or 0)
    support = float(sr.get('support') or 0)
    price = float(price)
    daily_log_g = float(daily_rate) / 100.0

    def sessions_to(target):
        if target <= 0 or price <= 0:
            return None
        # ln(target/price) / daily_log_g  (log-space sessions)
        s = np.log(target / price) / daily_log_g
        return int(round(s)) if s > 0 and np.isfinite(s) else None

    sessions_to_res = sessions_to(resistance)
    sessions_to_sup = sessions_to(support)

    return {
        "sessionsToResistance": sessions_to_res,
        "sessionsToSupport": sessions_to_sup,
        "resistanceDistancePct": round(((resistance - price) / price) * 100.0, 2),
        "supportDistancePct": round(((price - support) / price) * 100.0, 2),
        "trendQuality": projection.get('trendQuality'),
        "rSquared": projection.get('rSquared'),
        "direction": projection.get('direction'),
    }

def compute_support_resistance(df, lookback=50):
    """
    Support & resistance from the rolling high/low (excluding the very latest
    candle so the levels are meaningful) plus classic pivot points. Also reports
    where the price sits between support and resistance and how close it is to
    either level. Returns a dict for the latest bar.
    """
    if df is None or len(df) < 10 or 'high' not in df.columns or 'low' not in df.columns:
        return None
    data = df.reset_index(drop=True)
    last = int(len(data) - 1)
    win_start = max(0, last - lookback)
    window = data.iloc[win_start:last]
    if len(window) < 5:
        window = data.iloc[:last]

    resistance = float(window['high'].max())
    support = float(window['low'].min())

    prev_high = float(data.iloc[last - 1]['high'])
    prev_low = float(data.iloc[last - 1]['low'])
    prev_close = float(data.iloc[last - 1]['close'])
    pivot = (prev_high + prev_low + prev_close) / 3.0

    price = float(data.iloc[last]['close'])
    r1 = 2 * pivot - prev_low
    r2 = pivot + (prev_high - prev_low)
    s1 = 2 * pivot - prev_high
    s2 = pivot - (prev_high - prev_low)

    span = (resistance - support) if resistance > support else 1e-9
    position_pct = round(((price - support) / span) * 100.0, 1)
    break_range_pct = round((span / support) * 100.0 if support else 0.0, 2)

    return {
        "support": round(support, 2),
        "resistance": round(resistance, 2),
        "pivot": round(pivot, 2),
        "r1": round(r1, 2),
        "r2": round(r2, 2),
        "s1": round(s1, 2),
        "s2": round(s2, 2),
        "positionPct": position_pct,
        "rangePct": break_range_pct,
        "nearResistance": (resistance - price) <= 0.02 * price,
        "nearSupport": (price - support) <= 0.02 * price
    }

def project_trend(df, horizon_days=10, reg_lookback=30):
    """
    Project the expected move over the next `horizon_days` sessions using a log
    price linear-regression slope (trend) scaled by the regression R^2 (quality)
    and a volatility band from realised daily returns. Returns a dict with the
    direction, expected move %, a low/high band and a trend-quality label.
    """
    if df is None or len(df) < 20 or 'close' not in df.columns:
        return None
    closes = df['close'].dropna().astype(float)
    closes = closes[closes > 0].reset_index(drop=True)
    if len(closes) < 10:
        return None
    series = closes.iloc[-reg_lookback:]
    if len(series) < 10:
        series = closes

    log_p = np.log(series.to_numpy(dtype=float))
    x = np.arange(len(log_p), dtype=float)
    slope, intercept = np.polyfit(x, log_p, 1)
    y_hat = slope * x + intercept
    ss_res = float(np.sum((log_p - y_hat) ** 2))
    ss_tot = float(np.sum((log_p - np.mean(log_p)) ** 2)) or 1e-12
    r2 = float(np.clip(1.0 - ss_res / ss_tot, 0.0, 1.0))

    rets = np.diff(np.log(closes.to_numpy(dtype=float)))
    vol_daily = float(np.std(rets, ddof=1)) if len(rets) > 1 else 0.0
    z = 0.5

    daily_g = float(slope)
    drift = np.exp(daily_g * horizon_days) - 1.0
    band = vol_daily * np.sqrt(horizon_days) * z

    direction = "UP" if daily_g > 0 else ("DOWN" if daily_g < 0 else "FLAT")
    if r2 >= 0.5:
        quality = "strong"
    elif r2 >= 0.2:
        quality = "moderate"
    else:
        quality = "weak"

    return {
        "direction": direction,
        "horizonDays": horizon_days,
        "dailyRatePct": round(daily_g * 100.0, 4),
        "expectedMovePct": round(drift * 100.0, 2),
        "lowPct": round((drift - band) * 100.0, 2),
        "highPct": round((drift + band) * 100.0, 2),
        "rSquared": round(r2, 3),
        "trendQuality": quality,
        "volatilityDailyPct": round(vol_daily * 100.0, 3)
    }

if __name__ == "__main__":
    # Test calculation with dummy series
    dates = pd.date_range('2024-01-01', periods=60)
    sample_df = pd.DataFrame({
        'date': dates,
        'open': np.linspace(100, 150, 60),
        'high': np.linspace(102, 155, 60),
        'low': np.linspace(98, 148, 60),
        'close': np.linspace(101, 152, 60),
        'volume': np.random.randint(1000, 5000, 60)
    })
    res = compute_all_indicators(sample_df)
    print(f"Calculated indicators successfully. Result columns: {list(res.columns)}")

File: backend/src/ai/indicators.js
// Port of the AI engine's indicators.py (pandas/numpy) to plain JS.
// Input frames are arrays of {date, open, high, low, close, volume} sorted ascending.

function rsiSeries(closes, period = 14) {
  const n = closes.length;
  const gains = new Array(n).fill(0);
  const losses = new Array(n).fill(0);
  for (let i = 1; i < n; i++) {
    const d = closes[i] - closes[i - 1];
    gains[i] = Math.max(d, 0);
    losses[i] = Math.max(-d, 0);
  }
  // rolling(period, min_periods=1).mean()
  const out = new Array(n).fill(50);
  for (let i = 0; i < n; i++) {
    const start = Math.max(0, i - period + 1);
    let sg = 0, sl = 0;
    for (let j = start; j <= i; j++) { sg += gains[j]; sl += losses[j]; }
    const cnt = i - start + 1;
    const avgGain = sg / cnt;
    const avgLoss = sl / cnt;
    const rs = avgGain / (avgLoss || 1e-9);
    out[i] = 100 - 100 / (1 + rs);
  }
  return out;
}

function emaSeries(values, span) {
  const n = values.length;
  const out = new Array(n);
  const alpha = 2 / (span + 1);
  let prev = values[0];
  out[0] = prev;
  for (let i = 1; i < n; i++) {
    prev = values[i] * alpha + prev * (1 - alpha);
    out[i] = prev;
  }
  return out;
}

function smaSeries(values, period) {
  const n = values.length;
  const out = new Array(n).fill(0);
  let run = 0;
  for (let i = 0; i < n; i++) {
    run += values[i];
    if (i >= period) run -= values[i - period];
    const start = Math.max(0, i - period + 1);
    out[i] = run / (i - start + 1);
  }
  return out;
}

function rollingStd(values, period) {
  const n = values.length;
  const out = new Array(n).fill(0);
  for (let i = 0; i < n; i++) {
    const start = Math.max(0, i - period + 1);
    const cnt = i - start + 1;
    if (cnt < 2) continue;
    let sum = 0;
    for (let j = start; j <= i; j++) sum += values[j];
    const mean = sum / cnt;
    let ss = 0;
    for (let j = start; j <= i; j++) { const d = values[j] - mean; ss += d * d; }
    out[i] = Math.sqrt(ss / (cnt - 1));
  }
  return out;
}

function atrSeries(frame, period = 14) {
  const n = frame.length;
  const tr = new Array(n).fill(0);
  for (let i = 0; i < n; i++) {
    const hl = frame[i].high - frame[i].low;
    if (i === 0) { tr[i] = hl; continue; }
    const hc = Math.abs(frame[i].high - frame[i - 1].close);
    const lc = Math.abs(frame[i].low - frame[i - 1].close);
    tr[i] = Math.max(hl, hc, lc);
  }
  return smaSeries(tr, period);
}

function computeAllIndicators(frame) {
  const n = frame.length;
  const closes = frame.map(r => r.close);
  const rsi = rsiSeries(closes, 14);
  const emaFast = emaSeries(closes, 12);
  const emaSlow = emaSeries(closes, 26);
  const macd = emaFast.map((v, i) => v - emaSlow[i]);
  const macdSignal = emaSeries(macd, 9);
  const macdHist = macd.map((v, i) => v - macdSignal[i]);
  const sma20 = smaSeries(closes, 20);
  const sma50 = smaSeries(closes, 50);
  const ema20 = emaSeries(closes, 20);
  const ema50 = emaSeries(closes, 50);
  const std20 = rollingStd(closes, 20);
  const bbUpper = sma20.map((v, i) => v + std20[i] * 2);
  const bbLower = sma20.map((v, i) => v - std20[i] * 2);
  const bbMiddle = sma20.slice();
  const bbWidth = sma20.map((v, i) => (bbUpper[i] - bbLower[i]) / (v || 1e-9));
  const atr = atrSeries(frame, 14);

  const momentum5d = new Array(n).fill(0);
  if (n > 5) {
    const firstValid = (closes[5] - closes[0]) / closes[0] * 100;
    for (let i = 0; i < n; i++) {
      if (i >= 5 && closes[i - 5] > 0) {
        momentum5d[i] = (closes[i] - closes[i - 5]) / closes[i - 5] * 100;
      } else {
        momentum5d[i] = firstValid;
      }
    }
  }

  const out = new Array(n);
  for (let i = 0; i < n; i++) {
    const r = frame[i];
    out[i] = {
      date: r.date,
      open: r.open,
      high: r.high,
      low: r.low,
      close: r.close,
      volume: r.volume,
      rsi: numOr0(rsi[i]),
      macd: numOr0(macd[i]),
      macd_signal: numOr0(macdSignal[i]),
      macd_hist: numOr0(macdHist[i]),
      sma_20: numOr0(sma20[i]),
      sma_50: numOr0(sma50[i]),
      ema_20: numOr0(ema20[i]),
      ema_50: numOr0(ema50[i]),
      bb_upper: numOr0(bbUpper[i]),
      bb_middle: numOr0(bbMiddle[i]),
      bb_lower: numOr0(bbLower[i]),
      bb_width: numOr0(bbWidth[i]),
      atr: numOr0(atr[i]),
      volatility_pct: r.close > 0 ? numOr0((atr[i] / r.close) * 100) : 0,
      momentum_5d: numOr0(momentum5d[i]),
    };
  }
  return out;
}

function numOr0(v) {
  return isFinite(v) ? v : 0;
}

function toNum(v) {
  try { return v == null ? null : parseFloat(v); } catch (err) { return null; }
}

// Interpret the latest bar in plain words (mirrors indicator_signals).
function indicatorSignals(rows) {
  if (!rows || rows.length < 3) return null;
  const latest = rows[rows.length - 1];
  const close = toNum(latest.close) || 0;

  const rsi = toNum(latest.rsi);
  let rsiState = null;
  if (rsi != null) rsiState = rsi >= 70 ? 'OVERBOUGHT' : rsi <= 30 ? 'OVERSOLD' : 'NEUTRAL';

  const macd = toNum(latest.macd);
  const macdSignal = toNum(latest.macd_signal);
  let macdCross = null;
  let macdState = null;
  if (macd != null && macdSignal != null) {
    macdState = macd >= macdSignal ? 'ABOVE' : 'BELOW';
    const histTail = rows.slice(-6).map(r => toNum(r.macd_hist) || 0);
    for (let i = histTail.length - 1; i > 0; i--) {
      if (histTail[i - 1] <= 0 && histTail[i] > 0) { macdCross = 'BULLISH'; break; }
      if (histTail[i - 1] >= 0 && histTail[i] < 0) { macdCross = 'BEARISH'; break; }
    }
  }

  const sma20 = toNum(latest.sma_20);
  const sma50 = toNum(latest.sma_50);
  const ema20 = toNum(latest.ema_20);
  const ema50 = toNum(latest.ema_50);
  let vsSma20 = null, vsSma50 = null, smaTrend = null;
  if (sma20) vsSma20 = Math.round(((close - sma20) / sma20) * 10000) / 100;
  if (sma50) vsSma50 = Math.round(((close - sma50) / sma50) * 10000) / 100;
  if (sma20 && sma50) smaTrend = sma20 > sma50 ? 'UPTREND' : sma20 < sma50 ? 'DOWNTREND' : 'SIDEWAYS';

  const bbUpper = toNum(latest.bb_upper);
  const bbLower = toNum(latest.bb_lower);
  let bbPosition = null, bbState = null;
  if (bbUpper != null && bbLower != null && bbUpper > bbLower) {
    bbPosition = Math.round(((close - bbLower) / (bbUpper - bbLower)) * 1000) / 10;
    bbState = bbPosition >= 90 ? 'AT_UPPER' : bbPosition >= 60 ? 'UPPER_HALF' : bbPosition <= 10 ? 'AT_LOWER' : 'LOWER_HALF';
  }

  return {
    rsi: rsi != null ? Math.round(rsi * 10) / 10 : null,
    rsiState,
    macd: macd != null ? Math.round(macd * 10000) / 10000 : null,
    macdSignal: macdSignal != null ? Math.round(macdSignal * 10000) / 10000 : null,
    macdHist: macd != null && macdSignal != null ? Math.round((macd - macdSignal) * 10000) / 10000 : null,
    macdCross,
    macdState,
    priceVsSma20: vsSma20,
    priceVsSma50: vsSma50,
    ema20: ema20 != null ? Math.round(ema20 * 100) / 100 : null,
    ema50: ema50 != null ? Math.round(ema50 * 100) / 100 : null,
    smaTrend,
    bbPosition,
    bbState,
    atr: Math.round((toNum(latest.atr) || 0) * 100) / 100,
    volatilityPct: Math.round((toNum(latest.volatility_pct) || 0) * 100) / 100,
    momentum5d: Math.round((toNum(latest.momentum_5d) || 0) * 100) / 100,
  };
}

// Support & resistance from the rolling high/low + classic pivots.
function computeSupportResistance(frame, lookback = 50) {
  if (!frame || frame.length < 10) return null;
  const last = frame.length - 1;
  const start = Math.max(0, last - lookback);
  let window = frame.slice(start, last);
  if (window.length < 5) window = frame.slice(0, last);
  if (!window.length) return null;

  let resistance = -Infinity, support = Infinity;
  for (const r of window) {
    if (r.high > resistance) resistance = r.high;
    if (r.low < support) support = r.low;
  }
  const prev = frame[last - 1];
  const pivot = (prev.high + prev.low + prev.close) / 3;
  const price = frame[last].close;
  const r1 = 2 * pivot - prev.low;
  const r2 = pivot + (prev.high - prev.low);
  const s1 = 2 * pivot - prev.high;
  const s2 = pivot - (prev.high - prev.low);

  const span = resistance > support ? resistance - support : 1e-9;
  const positionPct = Math.round(((price - support) / span) * 1000) / 10;
  const rangePct = support ? Math.round((span / support) * 10000) / 100 : 0;

  return {
    support: Math.round(support * 100) / 100,
    resistance: Math.round(resistance * 100) / 100,
    pivot: Math.round(pivot * 100) / 100,
    r1: Math.round(r1 * 100) / 100,
    r2: Math.round(r2 * 100) / 100,
    s1: Math.round(s1 * 100) / 100,
    s2: Math.round(s2 * 100) / 100,
    positionPct,
    rangePct,
    nearResistance: (resistance - price) <= 0.02 * price,
    nearSupport: (price - support) <= 0.02 * price,
  };
}

// Least-squares log-price regression → projected move + band.
function projectTrend(frame, horizonDays = 10, regLookback = 30) {
  if (!frame || frame.length < 20) return null;
  const closes = frame.map(r => r.close).filter(c => c > 0);
  if (closes.length < 10) return null;
  let series = closes.slice(-regLookback);
  if (series.length < 10) series = closes;

  const m = series.length;
  const logP = series.map(v => Math.log(v));
  const x = Array.from({ length: m }, (_, i) => i);
  const xMean = x.reduce((a, b) => a + b, 0) / m;
  const yMean = logP.reduce((a, b) => a + b, 0) / m;
  let num = 0, den = 0;
  for (let i = 0; i < m; i++) { num += (x[i] - xMean) * (logP[i] - yMean); den += (x[i] - xMean) * (x[i] - xMean); }
  const slope = den > 0 ? num / den : 0;
  const intercept = yMean - slope * xMean;

  let ssRes = 0, ssTot = 0;
  for (let i = 0; i < m; i++) {
    const yhat = slope * x[i] + intercept;
    ssRes += (logP[i] - yhat) * (logP[i] - yhat);
    ssTot += (logP[i] - yMean) * (logP[i] - yMean);
  }
  const r2 = Math.max(0, Math.min(1, 1 - ssRes / (ssTot || 1e-12)));

  const logReturns = [];
  for (let i = 1; i < closes.length; i++) logReturns.push(Math.log(closes[i] / closes[i - 1]));
  const volMean = logReturns.reduce((a, b) => a + b, 0) / Math.max(1, logReturns.length);
  let volSq = 0;
  for (const r of logReturns) volSq += (r - volMean) * (r - volMean);
  const volDaily = logReturns.length > 1 ? Math.sqrt(volSq / (logReturns.length - 1)) : 0;
  const z = 0.5;

  const dailyG = slope;
  const drift = Math.exp(dailyG * horizonDays) - 1;
  const band = volDaily * Math.sqrt(horizonDays) * z;

  const direction = dailyG > 0 ? 'UP' : dailyG < 0 ? 'DOWN' : 'FLAT';
  const quality = r2 >= 0.5 ? 'strong' : r2 >= 0.2 ? 'moderate' : 'weak';

  return {
    direction,
    horizonDays,
    dailyRatePct: Math.round(dailyG * 100000) / 10000,
    expectedMovePct: Math.round(drift * 10000) / 100,
    lowPct: Math.round((drift - band) * 10000) / 100,
    highPct: Math.round((drift + band) * 10000) / 100,
    rSquared: Math.round(r2 * 1000) / 1000,
    trendQuality: quality,
    volatilityDailyPct: Math.round(volDaily * 100000) / 1000,
  };
}

function projectDuration(sr, projection, price) {
  if (!sr || !projection || price == null) return null;
  const dailyRate = projection.dailyRatePct;
  if (dailyRate == null || Math.abs(parseFloat(dailyRate)) < 1e-6) return null;

  const resistance = parseFloat(sr.resistance || 0);
  const support = parseFloat(sr.support || 0);
  const p = parseFloat(price);
  const dailyLogG = parseFloat(dailyRate) / 100;

  function sessionsTo(target) {
    if (target <= 0 || p <= 0) return null;
    const s = Math.log(target / p) / dailyLogG;
    return s > 0 && isFinite(s) ? Math.round(s) : null;
  }

  return {
    sessionsToResistance: sessionsTo(resistance),
    sessionsToSupport: sessionsTo(support),
    resistanceDistancePct: Math.round(((resistance - p) / p) * 10000) / 100,
    supportDistancePct: Math.round(((p - support) / p) * 10000) / 100,
    trendQuality: projection.trendQuality,
    rSquared: projection.rSquared,
    direction: projection.direction,
  };
}

module.exports = {
  computeAllIndicators,
  indicatorSignals,
  computeSupportResistance,
  projectTrend,
  projectDuration,
  rsiSeries,
  emaSeries,
  smaSeries,
};
